src/Moq.Analyzers/CallbackSignatureShouldMatchMockedMethodAnalyzer.cs

```diff
@@ -58,31 +58,101 @@ private static void Analyze(SyntaxNodeAnalysisContext context)
+            ITypeSymbol? lambdaParameterTypeSymbol = lambdaParameterType.Type;
+            ITypeSymbol? mockedMethodTypeSymbol = mockedMethodArgumentType.Type;
+
+            if (lambdaParameterTypeSymbol is null || mockedMethodTypeSymbol is null)
+            {
+                continue;
+            }
 
-                TypeInfo mockedMethodArgumentType = context.SemanticModel.GetTypeInfo(mockedMethodArguments[argumentIndex].Expression, context.CancellationToken);
+            if (!HasUserDefinedConversion(context.SemanticModel, mockedMethodTypeSymbol, lambdaParameterTypeSymbol))
+            {
+                Diagnostic diagnostic = lambdaParameters[argumentIndex].GetLocation().CreateDiagnostic(Rule);
+                context.ReportDiagnostic(diagnostic);
+            }
+        }
+    }
 
-                string? mockedMethodTypeName = mockedMethodArgumentType.ConvertedType?.ToString();
-                string? lambdaParameterTypeName = lambdaParameterType.ConvertedType?.ToString();
+    private static bool HasUserDefinedConversion(SemanticModel semanticModel, ITypeSymbol source, ITypeSymbol destination)
+    {
+        // This condition checks whether a valid type conversion exists between the parameter in the mocked method
+        // and the corresponding parameter in the callback lambda expression.
+        //
+        // - `conversion.Exists` checks if there is any type conversion possible between the two types
+        //
+        // The second part ensures that the conversion is either:
+        // 1. an implicit conversion,
+        // 2. an identity conversion (meaning the types are exactly the same), or
+        // 3. an explicit conversion.
+        //
+        // If the conversion exists, and it is one of these types (implicit, identity, or explicit), the analyzer will
+        // skip the diagnostic check, as the callback parameter type is considered compatible with the mocked method's
+        // parameter type.
+        Conversion conversion = semanticModel.Compilation.ClassifyConversion(source, destination);
+        bool conversionExists = conversion.Exists && (conversion.IsImplicit || conversion.IsExplicit || conversion.IsIdentity);
+
+        switch (conversionExists)
+        {
+            case true:
+                return true;
 
-                if (!string.Equals(mockedMethodTypeName, lambdaParameterTypeName, StringComparison.Ordinal))
+            // Fall back to seeking an implicit conversion operator
+            // The Conversion type will not show a user defined implicit conversion, so we have to seek it out explicitly
+            case false:
                 {
-                    Diagnostic diagnostic = callbackLambda.ParameterList.CreateDiagnostic(Rule);
-                    context.ReportDiagnostic(diagnostic);
+                    foreach (ISymbol? member in destination.GetMembers())
```

Comment:
This doesn't feel right to me. `Conversion` has an `IsUserDefined` method on it, why doesn't that work?

Reply:
Not sure why that doesn't work. Given the following scenario

```csharp
new Mock<IFoo>().Setup(x => x.Do((CustomType)"42")).Returns((CustomType i) => true);

internal interface IFoo
{
    bool Do(CustomType custom);
}

public class CustomType
{
    public int Value { get; }

    public CustomType(int value)
    {
        Value = value;
    }

    // User-defined conversions
    public static implicit operator CustomType(int value)
    {
        return new CustomType(value);
    }

    public static explicit operator CustomType(string str)
    {
        return new CustomType(int.Parse(str));
    }
}
```

The property for `conversion.IsUserDefined` is false. The only things that are true is `Exists`, `IsIdentity` and `IsImplicit`.

In the case `new Mock<IFoo>().Setup(x => x.Do(42)).Returns((CustomType i) => true);` `conversion` has `Explicit`, `IsUserDefined` and `IsImplicit` true.